### ISSUE TITLE
`LocalExecutor::make_default` leaks memory via RC cycle

### ISSUE THREAD

**Issue by matklad** [DataDog/scipio#110]
The

>And by all I mean, seriously, even spawn_without_executor() failed [ASAN] and there's clearly nothing wrong with that one.

[comment](https://github.com/DataDog/scipio/pull/104#issuecomment-707750633) got me curious, so I decided to `miri` that test.

```
$ MIRI_NIGHTLY=nightly-$(curl -s https://rust-lang.github.io/rustup-components-history/x86_64-unknown-linux-gnu/miri)
$ rustup toolchain add $MIRI_NIGHTLY
$ rustup component add miri --toolchain $MIRI_NIGHTLY
$ export MIRIFLAGS="-Zmiri-disable-isolation"
$ cargo +$MIRI_NIGHTLY miri test -- spawn_without_executor
```

Sure enough, it found a memory leak:

```
   Compiling scipio v0.1.0 (/home/matklad/projects/scipio/scipio)
   Compiling examples v0.0.0 (/home/matklad/projects/scipio/examples)
    Finished test [unoptimized + debuginfo] target(s) in 0.94s
     Running target/x86_64-unknown-linux-gnu/debug/deps/scipio-c56c9e43674a62a8

running 1 test
test executor::test::spawn_without_executor ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 83 filtered out

The following memory was leaked: alloc632956 (Rust heap, size: 168, align: 8) {
    0x00 │ 01 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 │ ................
    0x10 │ 00 00 00 00 00 00 00 00 08 00 00 00 00 00 00 00 │ ................
    ........................................
    ........................................
alloc633455 (fn: std::intrinsics::drop_in_place::<[closure@scipio/src/executor.rs:455:17: 458:18]> - shim(Some([closure@scipio/src/executor.rs:455:17: 458:18])))
alloc633456 (fn: executor::LocalExecutor::init::{closure#0})
alloc633457 (fn: executor::LocalExecutor::init::{closure#0})
alloc633458 (fn: <[closure@scipio/src/executor.rs:455:17: 458:18] as std::ops::FnOnce<()>>::call_once - shim(vtable))

error: the evaluated program leaked memory

error: aborting due to previous error

error: test failed, to rerun pass '-p scipio --lib'
```

Poking around `closure@scipio/src/executor.rs:455:17: 458:18 ` revealed what seems to an Rc-cycle memory leak. This commit adds a minimal demonstration and a crude fix via weak pointers:

https://github.com/matklad/scipio/commit/9f2912fe544ab46a96196d92959736d9c6350d0e

**Comment by glommer** [comment]
never heard of miri, and the output you posted is insanely helpful, a lot more than the asan default stuff.
I can only hope you benefiting from scipio one way or another as much as we are benefiting from your contributions.

I'll take a look at all of your pending stuff tomorrow morning

**Comment by glommer** [comment]
I now found the time to look at your crude fix.

You're almost on point. Indeed it is one of those "obvious in retrospect" memory leak cases: a queue lives in a queue collection (`available_executors`), but the notification function that the scheduler calls needs access to the queue collection to insert the task queue into the active executors list so it can run again.

Weak pointers are indeed the way to go. The main leap from your proposed fix is that while you insert that Peeble and seem to believe that the very fact that we constructed the executor is wrong (as it adds 1 to the Pebble). It is perfectly fine to build the executor, and it is perfectly fine to run and upgrade that pointer too.

It should not be possible to have that `upgrade` to fail, though: that means that the executor should be gone, which means nobody should be able to schedule us in. If somehow we were scheduled in, that's a bug. (Or if the queues somehow were destructed before the executor, that is a nastier bug).

So I will commit a fix now just upgrading a weak pointer every time we schedule.

Awesome!

(btw, miri doesn't seem to be able to execute some of the code, it complains that it can't call `unshare` so could not run a lot of other tests =( )
